src/algorithms/polygon_visibility/naive.rs

```diff
@@ -0,0 +1,198 @@
+//      should be Ordering::Greater than a point from a right-leaning ray.
+//
+// Pseudo code:
+//   for vertex in poly.vertex {
+//     let left_intersections = ... use left-leaning SoS ray ...;
+//     let right_intersections = ... use right-leaning SoS ray ...;
+//     let left_closest = ...;
+//     let right_closest = ...;
+//     output.push( (SoS::CounterClockWise, left_closest));
+//     output.push( (SoS::ClockWise, right_closest));
+//   }
+//   output.sort_by(|a,b| origin.ccw_cmp_around(a,b).then(cmp by sos) );
+//   Polygon::new(output)
+//
+// Test cases:
+//
+//   Input            Output
+//
+//  /-----\ /----\  /-----\ /----\
+//  |     | |    |  |     | |    |
+//  |  x  \-/ /--/  |  x  \---\--/
+//  |         |     |         |
+//  \---------/     \---------/
+//
+//
+//  /-----\ /----\  /-----\ /----\
+//  |     | |    |  |     | |    |
+//  |  x  \-/    |  |  x  \------\
+//  |            |  |            |
+//  \------------/  \------------/
+//
+//  /---------\     /-\     /-\
+//  |   /-\   |     |  \   /  |
+//  |   \-/   |     |   \ /   |
+//  |    x    |     |    x    |
+//  \---------/     \---------/
+
+pub fn get_visibility_polygon<T>(point: &Point<T, 2>, polygon: &Polygon<T>) -> Option<Polygon<T>>
+where
+  T: PolygonScalar,
+{
+  //check whether the point is in the polygon or not
+  if polygon.locate(point) == PointLocation::Outside {
+    return Option::None;
+  }
+
+  let polygon_segments: Vec<LineSegment<T, 2>> = polygon
+    .points
+    .windows(2)
+    .map(|pair| {
+      LineSegment::new(
+        EndPoint::Inclusive(pair[0].clone()),
+        EndPoint::Inclusive(pair[1].clone()),
+      )
+    })
+    .collect();
+
+  let mut polygon_points = Vec::new();
+  for vertex in &polygon.points {
+    let right_sos = HalfLineSoS {
+      line: LineSoS::new_through(point.clone(), vertex.clone()),
+      lean: SoS::ClockWise,
+    };
+    let left_sos = HalfLineSoS {
+      line: LineSoS::new_through(point.clone(), vertex.clone()),
+      lean: SoS::CounterClockWise,
+    };
+
+    let mut right_intersections = Vec::new();
+    let mut left_intersections = Vec::new();
+
+    polygon_segments.iter().for_each(|segment| {
+      if right_sos.intersect(segment.as_ref()).is_some() {
+        right_intersections.push(get_intersection(&right_sos, segment));
+      }
+      if left_sos.intersect(segment.as_ref()).is_some() {
+        left_intersections.push(get_intersection(&left_sos, segment));
+      }
+    });
+
+    if let Some(point) = right_intersections
+      .iter()
+      .max_by(|curr, next| point.cmp_distance_to(curr, next))
+    {
+      polygon_points.push(point.clone());
+    };
+    if let Some(point) = left_intersections
+      .iter()
+      .max_by(|curr, next| point.cmp_distance_to(curr, next))
+    {
+      polygon_points.push(point.clone());
+    };
+  }
+  polygon_points.sort_by(|curr, nxt| point.ccw_cmp_around(curr, nxt));
```

Comment:
Sorting the output points is no longer necessary after we've sorted the input vertices.